### PR TITLE
Add Modules/HeatConduction/ThermalContact syntax

### LINKED ISSUE #17259: Prepare doco for syntax in BISON for Module/HeatConduction
## Reason
There is a PR for a `Module/HeatConduction` action syntax in BISON that needs some extra MOOSE side corrections.

## Design
Add `Module/HeatConduction/index.md`

Correct md links.

## Impact
Better doco ability for downstream apps

### PR DESCRIPTION
**Pull request by tophmatthews**
Closes #17259

### CODE REVIEW

**Comment by tophmatthews** [comment]
Looking for feedback here, @bwspenc, @dschwen, others...

This creates a new syntax for the `ThermalContact` block, for which BISON hijacks with its own custom thermal contact methods. I would like access only to the thermal contact actions present in the `heat_conduction` modules. Instead of doing a bunch of messy deprecation in BISON that takes for ever, this creates a `Modules/HeatConduction/ThermalContact` block syntax, much like the `Modules/TensorMechanics` syntax, to ensure the heat conduction action stays nice and clean.

I added one test to verify the new syntax and required doco changes to avoid unspecific links. It requires a BISON patch, which is going through testing now. The other pieces in this PR are moving the `ThermalContact` tests from `combined` to `heat_conduction`, where they really belong.

**Comment by tophmatthews** [comment]
bump @dschwen @bwspenc

**Comment by tophmatthews** [comment]
The necessary fixes are in bison at this point, so this is all good to go. Can we at least start talking about solutions here?

**Comment by dschwen** [comment]
Yeah that looks nice

File: modules/heat_conduction/src/base/HeatConductionApp.C [hunk 1]
```diff
@@ -56,6 +56,19 @@ associateSyntaxInner(Syntax & syntax, ActionFactory & /*action_factory*/)
...
+  registerSyntaxTask("ThermalContactAction", "Modules/HeatConduction/ThermalContact/*", "add_bc");
+  registerSyntaxTask(
+      "ThermalContactAction", "Modules/HeatConduction/ThermalContact/*", "add_dirac_kernel");
+  registerSyntaxTask(
+      "ThermalContactAction", "Modules/HeatConduction/ThermalContact/*", "add_material");
+  registerSyntaxTask("ThermalContactAction",
+                     "Modules/HeatConduction/ThermalContact/*",
+                     "add_secondary_flux_vector");
+
   registerSyntaxTask("RadiationTransferAction", "GrayDiffuseRadiation/*", "append_mesh_generator");
```
Comment (dschwen) [inline modules/heat_conduction/src/base/HeatConductionApp.C:72]: This action should be under `/Modules/HeatConduction` as well... But that's for another PR. @snschune

**Comment by tophmatthews** [comment]
@bwspenc?

**Comment by tophmatthews** [comment]
@dschwen, If I remember correctly, @bwspenc was wanting to try and use the `type` parameter here, similar to basic moose blocks. For example, `Modules/HeatConduction/ThermalContact/type=GapHeatTransfer` would load only the necessary things for `GapHeatTransfer`. I think the problem there is that when `type=GapHeatTransfer`, `ThermalContactAction` does a lot of other things behind the scenes. Further, `GapHeatTransfer` is a `BC` object, not a `ThermalContact` object, but not sure that matters too much.

It's worth thinking through how a reworking of the contact system would work, but I hope this PR is a step in the right direction. It may be better if the syntax was applied here was only for `Modules/HeatConduction/ThermalContact/*` and not for anything down the road like `Modules/HeatConduction/ThermalContact/GapHeatTransfer`, Or better yet, `Modules/HeatConduction/ThermalContact/mortar` or `Modules/HeatConduction/ThermalContact/quadrature`.

Maybe this should be `Modules/HeatConduction/ThermalContact/BC` or `Modules/HeatConduction/ThermalContact/kinematic` to leave space for other methods?

**Comment by dschwen** [comment]
`type` is an argument of `MooseObjectAction`. Adding a `MooseEnum` or - god forbid - `std::string` `type` parameter is causing all kinds of confusion. Let's just _not_ do it.

**Comment by tophmatthews** [comment]
So I'm liking adding another later to this syntax to make room for other methods. What should it be?

`Modules/HeatConduction/ThermalContact/BC`

`Modules/HeatConduction/ThermalContact/kinematic`

`Modules/HeatConduction/ThermalContact/penetrationaux`

`Modules/HeatConduction/ThermalContact/old_model_we_have_to_use_until`@lindsayad`finally_wraps_up_mortar`

**Comment by dschwen** [comment]
Yeah, @topher, that is pretty much along the lines what @bwspenc and I talked about.

**Comment by tophmatthews** [comment]
Whats the correct lingo though?

**Comment by tophmatthews** [comment]
Wrong @ too @dschwen 😬

**Comment by dschwen** [comment]
Sorry @topher for the random GitHub notification :-D, and sorry @tophmatthews for messing up your handle!

**Comment by tophmatthews** [comment]
So whats the name that makes the most sense @dschwen!

**Comment by dschwen** [comment]
> So whats the name that makes the most sense

Honestly "topher" is. What's "toph' supposed to be?

**Comment by tophmatthews** [comment]
> > So whats the name that makes the most sense
> 
> Honestly "topher" is. What's "toph' supposed to be?

Aie Chihuahua, you know I meant name of method @dschwen! Anyways, I made it `Modules/HeatConduction/ThermalContact/BC`. Maybe we can finally wrap this up.

**Comment by tophmatthews** [comment]
Can we wrap this up @dschwen @bwspenc ? The syntax is now `Modules/HeatConduction/ThermalContact/BC` which I think makes the most sense and allows for other methods, e.g. `Modules/HeatConduction/ThermalContact/Mortar` in the future.

File: modules/heat_conduction/test/tests/gap_heat_transfer_htonly/gap_heat_transfer_htonly_test.i [hunk 2]
```diff
@@ -153,9 +153,10 @@
     thermal_conductivity = 100000000.0
   [../]
   [./density]
-    type = Density
+    type = GenericConstantMaterial
```
Comment (dschwen) [inline modules/heat_conduction/test/tests/gap_heat_transfer_htonly/gap_heat_transfer_htonly_test.i:156]: What's the deal with this substitution?

Comment (dschwen) [reply modules/heat_conduction/test/tests/gap_heat_transfer_htonly/gap_heat_transfer_htonly_test.i:156]: Oh, I see, this allows you to move it out of combined and into heat_condution!

**Review by dschwen — APPROVED** [review]
Ok, this is nice.

File: modules/heat_conduction/test/tests/gap_heat_transfer_htonly/gap_heat_transfer_htonly_test.i (hunk 2 above)
Comment (tophmatthews) [reply modules/heat_conduction/test/tests/gap_heat_transfer_htonly/gap_heat_transfer_htonly_test.i:156]: Yup! exactly. There are a lot of these types of tests in combined, that I have been slowly trying to migrate back to their respective modules whenever density is a constant. There are not even more improperly placed tests now that solid mechanics is gone.

**Pull request merged**